File: Algorithms/Python3.x/213-House_Robber_II.py
# Runtime: 40 ms, faster than 47.61% of Python3 online submissions for House Robber II.
# Memory Usage: 13.8 MB, less than 5.56% of Python3 online submissions for House Robber II.
class Solution:
    def rob(self, nums: List[int]) -> int:
        if nums == []:
            return 0
        if len(nums) <= 2:
            return max(nums)
        return max(self.dp(nums[1:]), self.dp(nums[:-1]))

    def dp(self, nums):
        cur, pre = 0, 0
        for num in nums:
            temp = max(num + pre, cur)
            pre = cur
            cur = temp
        return cur


# Runtime: 28 ms, faster than 87.18% of Python3 online submissions for House Robber II.
# Memory Usage: 14.2 MB, less than 60.09% of Python3 online submissions for House Robber II.
class Solution:
    def rob(self, nums: List[int]) -> int:
        if nums == []:
            return 0
        if len(nums) <= 2:
            return max(nums)
        return max(self.dp(nums[1:]), self.dp(nums[:-1]))

    def dp(self, nums):
        cur, pre = 0, 0
        for num in nums:
            pre, cur = cur, max(num + pre, cur)
        return cur

File: Algorithms/C++/213-House_Robber_II.cpp
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for House Robber II.
// Memory Usage: 7.7 MB, less than 84.51% of C++ online submissions for House Robber II.
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.size() == 1) {
            return nums[0];
        }
        if (nums.size() == 2) {
            return max(nums[0], nums[1]);
        }
        return max(dp(nums, 0, nums.size() - 1), dp(nums, 1, nums.size()));
    }
    
    int dp(const vector<int>& nums, int start, int end) {
        int cur = 0,
            pre = 0,
            tmp;
        for (int i = start; i < end; ++i) {
            tmp = max(nums[i] + pre, cur);
            pre = cur;
            cur = tmp;
        }
        return cur;
    }
};
